python/lsst/daf/butler/registries/sqlPreFlight.py

```diff
@@ -513,6 +519,88 @@ def _columns(selectable, names):
         subquery = subquery.alias("ds" + dsType.name)
         return subquery
 
+    def _buildUserExpression(self, expression, dimensions):
+        """Generate SQLAlchemy clause from user expression.
+
+        Parameters
+        ----------
+        expression : `exprParser.Node`
+            Parsed user expression tree.
+        dimensions : `DimensionGraph`
+            All Dimensions included in this query.
+
+        Returns
+        -------
+        clause : `sqlalchemy.ClauseElement`
+            Expression to be used in WHERE clause.
+
+        Raises
+        ------
+        ValueError
+            Raised for unexpected tree element, e.g. unknown link name.
+        Exception
+            Any exception raised by SQLAlchemy is propagated to caller.
+        """
+        if isinstance(expression, (exprParser.StringLiteral, exprParser.NumericLiteral)):
```

Comment:
I thought about adding that facility to exprParser tree classes but decided that I want to keep that code free of extra dependencies like sqlalchemy - if we want to support few additional tree transformations it would not be right design to do all of that inside tree itself. You are right that this is a common problem and some languages have better support for implementing this sort of double-dispatch code. I'll see if I can add "better" visitor patter support for those classes if it does not make things too complicated.

Reply:
@TallJimbo, I have re-implemented double dispatch using visitor pattern, could you check latest commit d11f7e4 to see if that smells better now.